Mathlib/Analysis/NormedSpace/MultipliableUniformlyOn.lean

```diff
@@ -0,0 +1,151 @@
+/-
+Copyright (c) 2025 Chris Birkbeck. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Birkbeck
+-/
+import Mathlib.Analysis.CStarAlgebra.Classes
+import Mathlib.Analysis.NormedSpace.FunctionSeries
+import Mathlib.Analysis.SpecialFunctions.Log.Summable
+import Mathlib.Topology.Algebra.InfiniteSum.UniformOn
+import Mathlib.Topology.Algebra.IsUniformGroup.Order
+
+/-!
+# Uniform convergence of products of functions
+
+We gather some results about the uniform convergence of infinite products, in particular those of
+the form `∏' i, (1 + f i x)` for a sequence `f` of complex valued functions.
+-/
+
+open Filter Function Complex Finset Topology
+
+variable {α ι : Type*}
+
+section summable
+
+variable {f : ι → α → ℂ} {g : α → ℂ} {K : Set α} {p : Filter ι}
+
+lemma TendstoUniformlyOn.comp_cexp {f : ι → α → ℂ} {g : α → ℂ}
+    (hf : TendstoUniformlyOn f g p K) (hg : BddAbove <| (fun x ↦ (g x).re) '' K) :
+    TendstoUniformlyOn (cexp ∘ f ·) (cexp ∘ g) p K := by
+  obtain ⟨v, hv⟩ : ∃ v, ∀ x ∈ K, (g x).re ≤ v := hg.imp <| by simp [mem_upperBounds]
+  have : ∀ᶠ i in p, ∀ x ∈ K, (f i x).re ≤ v + 1 := hf.re.eventually_forall_le (lt_add_one v) hv
+  refine (UniformContinuousOn.cexp _).comp_tendstoUniformlyOn_eventually (by simpa) ?_ hf
+  exact fun a ha ↦ (hv a ha).trans (lt_add_one v).le
+
+lemma Summable.hasSumUniformlyOn_log_one_add {f : ι → α → ℂ}
+    {u : ι → ℝ} (hu : Summable u) (h : ∀ᶠ n in cofinite, ∀ x ∈ K, ‖f n x‖ ≤ u n) :
+    HasSumUniformlyOn (fun i x ↦ log (1 + f i x)) (fun x ↦ ∑' i, log (1 + f i x)) {K} := by
+  simp only [hasSumUniformlyOn_iff_tendstoUniformlyOn, Set.mem_singleton_iff, forall_eq]
+  apply tendstoUniformlyOn_tsum_of_cofinite_eventually <| hu.mul_left (3 / 2)
+  filter_upwards [h, hu.tendsto_cofinite_zero.eventually_le_const one_half_pos] with i hi hn' x hx
+    using (norm_log_one_add_half_le_self <| (hi x hx).trans hn').trans (by simpa using hi x hx)
+
+lemma Summable.tendstoUniformlyOn_tsum_nat_log_one_add {f : ℕ → α → ℂ}
+    {u : ℕ → ℝ} (hu : Summable u) (h : ∀ᶠ n in atTop, ∀ x ∈ K, ‖f n x‖ ≤ u n) :
+    TendstoUniformlyOn (fun n x ↦ ∑ i ∈ Finset.range n, log (1 + f i x))
+    (fun x ↦ ∑' i : ℕ, Complex.log (1 + f i x)) atTop K := by
+  rw [← Nat.cofinite_eq_atTop] at h
+  exact (hu.hasSumUniformlyOn_log_one_add h).tendstoUniformlyOn_finset_range rfl
+
+end summable
+
+section UniformlyOn
+
+/-Note this is false without hfn. -/
```

Comment:
Rather more detail needed, I think. Here's a suggestion; please reformat this so it fits in the line width, & maybe write similarly detailed docstrings for some of the other lemmas.

```suggestion
/- If `x ↦ ∑' i, log (f i x)` is uniformly convergent on `𝔖`, its sum has bounded-above real part on each set in `𝔖`, and the functions `f i x` have no zeroes, then  `∏' i, f i x` is uniformly convergent on `𝔖`. 

Note that the non-vanishing assumption is really needed here: if this assumption is dropped then one obtains a counterexample by taking `ι = α = ℕ` and `f i x` is 0 if `i = x` and `1` otherwise. -/
```